src/json-patch-duplex.ts

```diff
@@ -524,95 +513,263 @@ module jsonpatch {
             }
-            key = parseInt(key, 10);
+            key = ~~key;
           }
           if (t >= len) {
-            if (validate && patch.op === "add" && key > obj.length) {
-              throw new JsonPatchError("The specified index MUST NOT be greater than the number of elements in the array", "OPERATION_VALUE_OUT_OF_BOUNDS", p - 1, patch.path, patch);
+            if (validateOperation && operation.op === "add" && key > length) {
+              throw new JsonPatchError("The specified index MUST NOT be greater than the number of elements in the array", "OPERATION_VALUE_OUT_OF_BOUNDS", 0, operation.path, operation);
+            }
+            returnValue.result = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
+            returnValue.newDocument = document;
+            if(returnValue.result === false) {
+                throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
             }
-            results.push(arrOps[patch.op].call(patch, obj, key, tree)); // Apply patch
-            break;
+            return returnValue;
           }
         }
         else {
-          if (key && key.indexOf('~') != -1)
-            key = key.replace(/~1/g, '/').replace(/~0/g, '~'); // escape chars
+          if (key && key.indexOf('~') != -1) {
+            key = unescapePathComponent(key);
+          }
           if (t >= len) {
-            results.push(objOps[patch.op].call(patch, obj, key, tree)); // Apply patch
-            break;
+            returnValue.result = objOps[operation.op].call(operation, obj, key, document); // Apply patch
+            returnValue.newDocument = document;
+            if(returnValue.result === false) {
+                throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
+            }
+            return returnValue;
           }
         }
         obj = obj[key];
       }
     }
+  }
+
+  /**
+   * Apply a full JSON Patch array on a JSON document.
+   * Returns the {newDocument, result} of the patch.
+   *
+   * @param document The document to patch
+   * @param patch The patch to apply
+   * @param validateOperation `false` is without validation, `true` to use default jsonpatch's validation, or you can pass a `validateOperation` callback to be used for validation.
+   * @return An array of `{newDocument, result}` after the patch
+   */
+  export function applyPatch<T>(document: T, patch: Operation[], validateOperation?: boolean | Validator<T>): OperationResult<T>[] {
+    const results: OperationResult<any>[] = new Array(patch.length);
+
+    for (let i = 0, length = patch.length; i < length; i++) {
+      results[i] = applyOperation(document, patch[i], validateOperation);
+      document = results[i].newDocument; // in case root was replaced
+    }
     return results;
```

Comment:
Maybe we should attach `results.newDocument = document` as well, so you would be able to check just `const newDoc = jsonpatch.applyPatch(oldDoc, patch).newDocument`, without a need to check the length of a patch and do
`const newDoc = jsonpatch.applyPatch(oldDoc, patch)[patch.length].newDocument` but maybe that's not that bad, either way we need to update the doc or change behavior here.